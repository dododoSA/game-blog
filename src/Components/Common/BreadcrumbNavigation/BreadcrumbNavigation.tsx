import { useRouter } from 'next/router';
import Link from 'next/link';

// TODO: ちょっと面倒くさそうなので一旦後回し 後で直しておく
const BreadcrumbNavigation = () => {
  const router = useRouter();
  const paths = router.pathname.split(/(?=\/)/g);
  let tmp = '';
  const breadcrumbs = paths.map((p) => {
    tmp += p;
    return (
      <Link key={tmp} href={tmp}>
        <a>{tmp}</a>
      </Link>
    );
  });

  return <>{router.pathname !== '/' && <div>{breadcrumbs}</div>}</>;
};

export { BreadcrumbNavigation };
