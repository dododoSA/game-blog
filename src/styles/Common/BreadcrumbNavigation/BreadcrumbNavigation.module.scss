.container {
  padding: 15px 0;
  margin: 5px;

  ul {
    display: flex;
    list-style: none;
    flex-wrap: wrap;
    align-items: center;
    justify-content: left;
    padding: 0;
    font-size: 1.4rem;

    li:nth-child(n+2)::before {
      content: '>';
      margin: 0 10px;
    }

    li:nth-last-child(n+2) {
      color: #808080;
    }

    li > a:hover {
      text-decoration: underline;
      color: #1D3557;
    }
  }
}