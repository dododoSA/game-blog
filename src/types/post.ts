export interface IPost {
  slug?: string;
  category: string;
  subcategory1?: string;
  subcategory2?: string;
  date: string;
  thumbnail: string;
  title: string;
  description: string;
}
