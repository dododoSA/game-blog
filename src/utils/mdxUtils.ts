import fs from 'fs';
import { join } from 'path';
import matter from 'gray-matter';
import { IPost } from 'types/post';
import { BREADCRUMB_LABEL } from 'const/breadcrumbLabel';

const POSTS_PATH = join(process.cwd(), 'posts');

const getPostFilePaths = (MDX_PATH: string) =>
  fs.readdirSync(join(POSTS_PATH, MDX_PATH)).filter((path) => /\.mdx?$/.test(path));

const getPost = (slug: string, path: string) => {
  const fullPath = join(POSTS_PATH, path, `${slug}.mdx`);
  const fileContents = fs.readFileSync(fullPath, 'utf8');
  return matter(fileContents);
};

type Items = {
  [key: string]: string;
};

const getPostItems = (fileName: string, dirPath: string, fields: string[] = []) => {
  const slug = fileName.replace(/\.mdx?$/, '');
  const { data, content } = getPost(slug, dirPath);
  const items: Items = {};

  // Ensure only the minimal needed data is exposed TODO: 確認
  fields.forEach((field) => {
    if (field === 'content') {
      items[field] = content;
    }
    if (data[field]) {
      items[field] = data[field];
    }
  });

  return items;
};

const getAllPosts = (dirPath: string, fields: string[] = []) => {
  const filePaths = getPostFilePaths(dirPath);
  const posts = filePaths
    .map((filePath) => getPostItems(filePath, dirPath, fields))
    .sort((post1, post2) => (post1.date > post2.date ? -1 : 1));
  return posts;
};

const getBreadcrumbPaths = (
  path: string,
  frontMatter?: IPost,
): {
  href: string;
  label: string;
}[] => {
  const tmpPath = `//${path}`;
  const dividedPath = tmpPath.split(/(?=\/)/g);
  let href = '';
  const breadcrumbPaths = dividedPath.map((p, index) => {
    if (index !== 0) {
      href += p;
    }
    return {
      href: index === 0 ? '/' : href,
      // ディレクトリを自動で取得できない以上これは手動で確認するしかない
      label: BREADCRUMB_LABEL.get(p) as string,
    };
  });

  if (frontMatter) {
    breadcrumbPaths.push({
      href: `${href}/${frontMatter.slug}`,
      label: frontMatter.title,
    });
  }

  return breadcrumbPaths;
};

export { getPost, getAllPosts, getBreadcrumbPaths };
