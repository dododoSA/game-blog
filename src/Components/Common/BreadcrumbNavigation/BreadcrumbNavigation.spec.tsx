import renderer from 'react-test-renderer';
import { BreadcrumbNavigation } from 'Components/Common/BreadcrumbNavigation/BreadcrumbNavigation';
import { getBreadcrumbPaths, getPost } from 'utils/mdxUtils';
import { IPost } from 'types/post';

jest.mock('next/router', () => ({
  useRouter() {
    return {
      // '/' 意外ならなんでも
      pathname: '/theory/pglanguage/c/setup',
    };
  },
}));

test('BreadcrumbNavigation', () => {
  // TODO: いろんなパスで正しいかどうかをチェックしたほうがいいかを考える
  const MDX_PATH = 'theory/pglanguage/c';
  const { data } = getPost('setup', MDX_PATH);
  const breadcrumbPaths = getBreadcrumbPaths(MDX_PATH, data as IPost);
  const component = renderer.create(<BreadcrumbNavigation paths={breadcrumbPaths} />);
  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
