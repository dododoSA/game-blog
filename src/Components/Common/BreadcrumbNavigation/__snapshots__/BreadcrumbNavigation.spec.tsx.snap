// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbNavigation 1`] = `
<div
  className="container"
>
  <ul>
    <li>
      <a
        href="/"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        HOME
      </a>
    </li>
    <li>
      <a
        href="/theory"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        理論編
      </a>
    </li>
    <li>
      <a
        href="/theory/pglanguage"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        プログラミング言語
      </a>
    </li>
    <li>
      <a
        href="/theory/pglanguage/c"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        C言語
      </a>
    </li>
    <li>
      <a
        href="/theory/pglanguage/c/setup"
        onClick={[Function]}
        onMouseEnter={[Function]}
      >
        事前準備
      </a>
    </li>
  </ul>
</div>
`;
